File: src/resolver/graph.cpp
#include "graph.hpp"
#include <algorithm>
#include <stack>

using namespace std;

DependencyGraph::DependencyGraph() {}

void DependencyGraph::addPackage(const Package& package) {
    string name = package.getName();
    packages[name] = package;
    
    // Initialize adjacency lists if needed
    if (adj_list.find(name) == adj_list.end()) {
        adj_list[name] = vector<string>();
    }
    if (reverse_adj_list.find(name) == reverse_adj_list.end()) {
        reverse_adj_list[name] = vector<string>();
    }
}

void DependencyGraph::addDependency(const string& from, const string& to) {
    // Add to forward adjacency list
    if (adj_list.find(from) == adj_list.end()) {
        adj_list[from] = vector<string>();
    }
    adj_list[from].push_back(to);
    
    // Add to reverse adjacency list
    if (reverse_adj_list.find(to) == reverse_adj_list.end()) {
        reverse_adj_list[to] = vector<string>();
    }
    reverse_adj_list[to].push_back(from);
}

vector<string> DependencyGraph::getPackages() const {
    vector<string> result;
    for (const auto& pair : packages) {
        result.push_back(pair.first);
    }
    return result;
}

vector<string> DependencyGraph::getDependencies(const string& package) const {
    auto it = adj_list.find(package);
    if (it != adj_list.end()) {
        return it->second;
    }
    return vector<string>();
}

vector<string> DependencyGraph::getDependents(const string& package) const {
    auto it = reverse_adj_list.find(package);
    if (it != reverse_adj_list.end()) {
        return it->second;
    }
    return vector<string>();
}

bool DependencyGraph::hasPackage(const string& package) const {
    return packages.find(package) != packages.end();
}

optional<Package> DependencyGraph::getPackage(const string& name) const {
    auto it = packages.find(name);
    if (it != packages.end()) {
        return it->second;
    }
    return nullopt;
}

vector<string> DependencyGraph::topologicalSort() const {
    vector<string> result;
    set<string> visited;
    
    // DFS-based topological sort
    for (const auto& pair : packages) {
        if (visited.find(pair.first) == visited.end()) {
            dfsTopological(pair.first, visited, result);
        }
    }
    
    // Reverse to get correct order (dependencies first)
    reverse(result.begin(), result.end());
    return result;
}

void DependencyGraph::dfsTopological(const string& node, set<string>& visited, vector<string>& result) const {
    visited.insert(node);
    
    // Visit all dependencies first
    auto deps = getDependencies(node);
    for (const auto& dep : deps) {
        if (visited.find(dep) == visited.end()) {
            dfsTopological(dep, visited, result);
        }
    }
    
    // Add node after dependencies
    result.push_back(node);
}

bool DependencyGraph::hasCycle() const {
    map<string, Color> colors;
    
    // Initialize all nodes as white
    for (const auto& pair : packages) {
        colors[pair.first] = Color::WHITE;
    }
    
    // Check each node
    for (const auto& pair : packages) {
        if (colors[pair.first] == Color::WHITE) {
            vector<string> cycle;
            if (dfsCycle(pair.first, colors, cycle)) {
                return true;
            }
        }
    }
    
    return false;
}

vector<string> DependencyGraph::getCycle() const {
    map<string, Color> colors;
    vector<string> cycle;
    
    // Initialize all nodes as white
    for (const auto& pair : packages) {
        colors[pair.first] = Color::WHITE;
    }
    
    // Find cycle
    for (const auto& pair : packages) {
        if (colors[pair.first] == Color::WHITE) {
            if (dfsCycle(pair.first, colors, cycle)) {
                return cycle;
            }
        }
    }
    
    return vector<string>();
}

bool DependencyGraph::dfsCycle(const string& node, map<string, Color>& colors, vector<string>& cycle) const {
    colors[node] = Color::GRAY;
    cycle.push_back(node);
    
    auto deps = getDependencies(node);
    for (const auto& dep : deps) {
        if (colors[dep] == Color::GRAY) {
            // Found back edge - cycle detected
            return true;
        } else if (colors[dep] == Color::WHITE) {
            if (dfsCycle(dep, colors, cycle)) {
                return true;
            }
        }
    }
    
    colors[node] = Color::BLACK;
    cycle.pop_back();
    return false;
}

vector<vector<string>> DependencyGraph::getStronglyConnectedComponents() const {
    map<string, int> index;
    map<string, int> lowlink;
    vector<string> stack;
    map<string, bool> on_stack;
    int current_index = 0;
    vector<vector<string>> sccs;
    
    // Initialize
    for (const auto& pair : packages) {
        index[pair.first] = -1;
        lowlink[pair.first] = -1;
        on_stack[pair.first] = false;
    }
    
    // Run Tarjan's algorithm
    for (const auto& pair : packages) {
        if (index[pair.first] == -1) {
            tarjanDFS(pair.first, index, lowlink, stack, on_stack, current_index, sccs);
        }
    }
    
    return sccs;
}

void DependencyGraph::tarjanDFS(const string& node,
                                map<string, int>& index,
                                map<string, int>& lowlink,
                                vector<string>& stack,
                                map<string, bool>& on_stack,
                                int& current_index,
                                vector<vector<string>>& sccs) const {
    index[node] = current_index;
    lowlink[node] = current_index;
    current_index++;
    stack.push_back(node);
    on_stack[node] = true;
    
    auto deps = getDependencies(node);
    for (const auto& dep : deps) {
        if (index[dep] == -1) {
            tarjanDFS(dep, index, lowlink, stack, on_stack, current_index, sccs);
            lowlink[node] = min(lowlink[node], lowlink[dep]);
        } else if (on_stack[dep]) {
            lowlink[node] = min(lowlink[node], index[dep]);
        }
    }
    
    if (lowlink[node] == index[node]) {
        vector<string> scc;
        string w;
        do {
            w = stack.back();
            stack.pop_back();
            on_stack[w] = false;
            scc.push_back(w);
        } while (w != node);
        sccs.push_back(scc);
    }
}

void DependencyGraph::clear() {
    adj_list.clear();
    reverse_adj_list.clear();
    packages.clear();
}

File: dpm/resolver/graph.py
"""
graph.py - dependency graph for managing package relationships
"""

from typing import Dict, List, Set, Optional
from ..core.package import Package


class DependencyGraph:
    """dependency graph for managing package relationships"""
    
    def __init__(self):
        self.packages: Dict[str, Package] = {}  # package name -> package
        self.dependencies: Dict[str, List[str]] = {}  # package -> list of dependencies
        self.dependents: Dict[str, List[str]] = {}  # package -> list of packages that depend on it
    
    def add_package(self, package: Package):
        """add a package to the graph"""
        self.packages[package.name] = package
        if package.name not in self.dependencies:
            self.dependencies[package.name] = []
        if package.name not in self.dependents:
            self.dependents[package.name] = []
    
    def add_dependency(self, package: str, dependency: str):
        """add a dependency edge: package depends on dependency"""
        if package not in self.dependencies:
            self.dependencies[package] = []
        if dependency not in self.dependencies[package]:
            self.dependencies[package].append(dependency)
        
        # update dependents
        if dependency not in self.dependents:
            self.dependents[dependency] = []
        if package not in self.dependents[dependency]:
            self.dependents[dependency].append(package)
    
    def get_package(self, name: str) -> Optional[Package]:
        """get a package by name"""
        return self.packages.get(name)
    
    def get_dependencies(self, name: str) -> List[str]:
        """get dependencies of a package"""
        return self.dependencies.get(name, [])
    
    def get_dependents(self, name: str) -> List[str]:
        """get packages that depend on this one"""
        return self.dependents.get(name, [])
    
    def get_packages(self) -> List[str]:
        """get all package names"""
        return list(self.packages.keys())
    
    def has_cycle(self) -> bool:
        """check if the graph has a cycle"""
        return len(self._find_cycle()) > 0
    
    def get_cycle(self) -> List[str]:
        """get a cycle if one exists"""
        return self._find_cycle()
    
    def _find_cycle(self) -> List[str]:
        """find a cycle using dfs"""
        visited: Set[str] = set()
        rec_stack: Set[str] = set()
        cycle_path: List[str] = []
        
        def dfs(node: str, path: List[str]) -> bool:
            visited.add(node)
            rec_stack.add(node)
            path.append(node)
            
            for neighbor in self.dependencies.get(node, []):
                if neighbor not in visited:
                    if dfs(neighbor, path):
                        return True
                elif neighbor in rec_stack:
                    # found a cycle
                    cycle_start = path.index(neighbor)
                    cycle_path.extend(path[cycle_start:] + [neighbor])
                    return True
            
            rec_stack.remove(node)
            path.pop()
            return False
        
        for node in self.packages.keys():
            if node not in visited:
                if dfs(node, []):
                    return cycle_path
        
        return []
    
    def topological_sort(self) -> List[str]:
        """topological sort of packages (dependencies before dependents)"""
        # build reverse graph: what depends on what
        reverse_deps: Dict[str, List[str]] = {pkg: [] for pkg in self.packages.keys()}
        for pkg, deps in self.dependencies.items():
            for dep in deps:
                if dep in reverse_deps:
                    reverse_deps[dep].append(pkg)
        
        # calculate in-degrees (how many dependencies this package has)
        in_degree: Dict[str, int] = {}
        for pkg in self.packages.keys():
            in_degree[pkg] = len(self.dependencies.get(pkg, []))
        
        # kahn's algorithm - start with packages that have no dependencies
        queue = [pkg for pkg, degree in in_degree.items() if degree == 0]
        result = []
        
        while queue:
            node = queue.pop(0)
            result.append(node)
            
            # for each package that depends on this node, decrease its in-degree
            for dependent in reverse_deps.get(node, []):
                if dependent in in_degree:
                    in_degree[dependent] -= 1
                    if in_degree[dependent] == 0:
                        queue.append(dependent)
        
        # if we didn't process all nodes, there's a cycle
        if len(result) != len(self.packages):
            return []
        
        return result

